cr-examples/samples/src/main/java/oracle/code/samples/MathOptimizer.java

```diff
@@ -0,0 +1,281 @@
+ * 2) Replace Pow(x, y) when y == 2 to (x * x).
+ * </p>
+ *
+ * <p>
+ *     Babylon repository: {@see <a href="https://github.com/openjdk/babylon/tree/code-reflection">link</a>}
+ * </p>
+ *
+ * <p>
+ *     How to run?
+ *     <code>
+ *         java --enable-preview -cp target/crsamples-1.0-SNAPSHOT.jar oracle.code.samples.MathOptimizer
+ *     </code>
+ * </p>:
+ */
+public class MathOptimizer {
+
+    @CodeReflection
+    private static double myFunction(int value) {
+        return Math.pow(2, value);
+    }
+
+    // if pow(2, x), then substitute for this function
+    // We could apply this function if, at runtime, user pass int values to the pow function
+    // Thus, we narrow the result type from 8 bytes (double) to 4 bytes (INT).
+    private static int functionShift(int val) {
+        return 1 << val;
+    }
+
+    // if pow(x, 2) then substitute for this function
+    private static double functionMult(double x) {
+        return x * x;
+    }
+
+    private static final MethodRef MY_SHIFT_FUNCTION = MethodRef.method(MathOptimizer.class, "functionShift", int.class, int.class);
+
+    private static final MethodRef MY_MULT_FUNCTION = MethodRef.method(MathOptimizer.class, "functionMult", double.class, double.class);
+
+    // Analyze type methods: taken from example of String Concat Transformer to traverse the tree.
+    static boolean analyseType(JavaOp.ConvOp convOp, JavaType typeToMatch) {
+        return analyseType(convOp.operands().get(0), typeToMatch);
+    }
+
+    static boolean analyseType(Value v, JavaType typeToMatch) {
+        // Maybe there is an utility already to do tree traversal
+        if (v instanceof Op.Result r && r.op() instanceof JavaOp.ConvOp convOp) {
+            // Node of tree, recursively traverse the operands
+            return analyseType(convOp, typeToMatch);
+        } else {
+            // Leaf of tree: analyze type
+            TypeElement type = v.type();
+            return type.equals(typeToMatch);
+        }
+    }
+
+    static void main(String[] args) {
+
+        Optional<Method> myFunction = Stream.of(MathOptimizer.class.getDeclaredMethods())
+                .filter(m -> m.getName().equals("myFunction"))
+                .findFirst();
+
+        Method myMathMethod = myFunction.get();
+
+        // Obtain the code model for the annotated method
+        CoreOp.FuncOp codeModel = Op.ofMethod(myMathMethod).get();
+        System.out.println(codeModel.toText());
+
+        System.out.println("\nLet's transform the code");
+        codeModel = codeModel.transform(CopyContext.create(), (blockBuilder, op) -> {
+            switch (op) {
+                case JavaOp.InvokeOp invokeOp when whenIsMathPowFunction(invokeOp) -> {
+                    // The idea here is to create a new JavaOp.invoke with the optimization and replace it.
+                    List<Value> operands = blockBuilder.context().getValues(op.operands());
```

Comment:
Strictly speaking `operands` are the _output_ operands of values in output model we are building. We get them from the _input_ operands, values in the input model. This works because the transformation has already mapped those input values to output values, since it already encountered them (when copying of the func op and its parameters).